src/ui/ui_main_menu.c

```diff
@@ -349,39 +354,33 @@ void main_menu_init(void) {
     keyboard_init();
 }
 
+static void bootup_action_completed() {
+    bootup_actions_fired = false;
+}
+
+static void handle_bootup_action() {
+    static page_pack_t **next_bootup_action = &post_bootup_actions[0];
+    if (next_bootup_action - &post_bootup_actions[0] >= post_bootup_actions_count) {
+        return;
+    }
+
+    (*next_bootup_action++)->post_bootup_run_function(bootup_action_completed);
+}
+
 void main_menu_update() {
     static uint32_t delta_ms = 0;
-    static uint32_t last_bootup_action = 0;
     uint32_t now_ms = time_ms();
     delta_ms = now_ms - delta_ms;
 
     for (uint32_t i = 0; i < PAGE_COUNT; i++) {
         if (page_packs[i]->on_update) {
             page_packs[i]->on_update(delta_ms);
         }
+    }
 
-        if (last_bootup_action == i && post_bootup_actions[i]) {
-            if (post_bootup_actions[i] != NULL) {
-                // Function invokation
-                if (post_bootup_actions[i]->post_bootup_run_complete == NULL) {
-                    if (post_bootup_actions[i]->post_bootup_run_function != NULL) {
-                        post_bootup_actions[i]->post_bootup_run_function();
-                        post_bootup_actions[i]->post_bootup_run_function = NULL;
-                        post_bootup_actions[i] = NULL;
-                        ++last_bootup_action;
-                    }
-                } else { // Thread invokation
-                    if (post_bootup_actions[i]->post_bootup_run_function){
-                        post_bootup_actions[i]->post_bootup_run_function();
-                        post_bootup_actions[i]->post_bootup_run_function = NULL;
-                    } else if (post_bootup_actions[i]->post_bootup_run_complete()) {
-                        post_bootup_actions[i]->post_bootup_run_complete = NULL;
-                        post_bootup_actions[i] = NULL;
-                        ++last_bootup_action;
-                    }
-                }
-            }
-        }
+    if (!bootup_actions_fired) {
+        bootup_actions_fired = true;
+        handle_bootup_action();
```

Comment:
Here you can pass bootup_actions_fired to handle_bootup_action(&bootup_actions_fired) and continue to pass the boolean all the way to 

(*next_bootup_action++)->post_bootup_run_function(bootup_actions_fired)